internal/controller/sync.go

```diff
@@ -63,47 +68,58 @@ func (r *Route) sync(ctx context.Context, req reconcile.Request, route *routev1.
 	if r.hasValidCertificate(route) {
 		result, err = reconcile.Result{RequeueAfter: r.getRequeueAfterDuration(route)}, nil
 		log.V(5).Info("route has valid cert")
-		return
+		return result, err
 	}
 	// Do we have a revision? If not set revision to 0
 	revision, err := getCurrentRevision(route)
 	if err != nil {
 		err = r.setRevision(ctx, route, 0)
 		log.V(5).Info("generated revision 0")
-		return
+		return result, err
 	}
 	// Do we have a next key?
 	if !r.hasNextPrivateKey(route) {
 		err = r.generateNextPrivateKey(ctx, route)
 		log.V(5).Info("generated next private key for route")
-		return
+		return result, err
 	}
 	// Is there a CertificateRequest for the Next revision? If not, make it.
 	hasNext, err := r.hasNextCR(ctx, route, revision)
 	if err != nil {
-		// err above is the returned err - named returns parameters + bare returns can be confusing
-		return
+		return result, err
 	}
 	if !hasNext {
-		// create CR and return. We own the CR so it will cause a re-reconcile
+		// generate manifest for new CR
 		log.V(5).Info("route has no matching certificate request", "revision", revision)
-		err = r.createNextCR(ctx, route, revision)
-		return
+		var cr *cmapi.CertificateRequest
+		cr, err = r.buildNextCR(ctx, route, revision)
+		if err != nil {
+			log.V(1).Error(err, "error generating certificate request", "object", req.NamespacedName)
+			// Not a reconcile error, so don't retry this revision
+			return result, nil
+		}
+
+		// create CR and return. We own the CR so it will cause a re-reconcile
+		_, err = r.certClient.CertmanagerV1().CertificateRequests(route.Namespace).Create(ctx, cr, metav1.CreateOptions{})
+		if err != nil {
+			return result, err
+		}
+		r.eventRecorder.Event(route, corev1.EventTypeNormal, ReasonIssuing, "Created new CertificateRequest for Route %s")
```

Comment:
Shouldn't we return right after creating the CR? The CR will trigger a reconciliation of the Route, so we can stop now, no?

Reply:
You're totally right! This got lost in the refactor. Fixed now.